server/src/main/java/org/elasticsearch/common/logging/HeaderWarningAppender.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.logging;
+
+import org.apache.logging.log4j.core.Appender;
+import org.apache.logging.log4j.core.Core;
+import org.apache.logging.log4j.core.Filter;
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.core.appender.AbstractAppender;
+import org.apache.logging.log4j.core.config.plugins.Plugin;
+import org.apache.logging.log4j.core.config.plugins.PluginAttribute;
+import org.apache.logging.log4j.core.config.plugins.PluginElement;
+import org.apache.logging.log4j.core.config.plugins.PluginFactory;
+import org.apache.logging.log4j.message.Message;
+
+@Plugin(name = "HeaderWarningAppender", category = Core.CATEGORY_NAME, elementType = Appender.ELEMENT_TYPE)
```

Comment:
I haven't changed `HeaderWarning` though, I've added an appender on top of it. I see the argument that's it's not a 'thing' to write to, but another way of looking at appenders is that they're just 'sinks' for log events. log4j ships with appenders that write to message queues, for example.

We can still add headers using `HeaderWarning`, or configure the appender for other messages as well.

Reply:
I wonder maybe if we decide to keep `HeaderWarningAppender` then maybe we should create a logger for it and limit the scope of `HeaderWarning`? To make sure that there is single and consistent way of emitting header warnings.

Reply:
@pugnascotia any views on creating a logger?

Reply:
@pgomulka I'm not clear what you are suggesting - are you thinking of adding something like `DeprecationLogger` but for header warnings?

Reply:
I was thinking of configuring a logger that uses a `header_warning` appender. Named maybe `header_logger` ? and the usage in https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java#L483
would be something like
```
logger = LogManger.getLogger("header_logger");
logger.warn(warning);
```

It worries me that there are two ways to emit those warnings - by log4j and directly.

Reply:
Personally, I don't find the current implementation too troubling, but then I was thinking about what the code looks like when we have more places that want to emit warnings back to the client, and even possibly record those to an index. In that world, it starts to make a lot of sense to separate the loggers, so I think I will do as you suggest.

Reply:
@pgomulka I've implemented a `header_logger` in 76993c95f96.

Reply:
I'm failing to understand what the *advantage* is of using log4j for header warnings? In either case, the warnings are stored in the thread context still right? Is it for formatting?

If we are going to use it, then I think it should be hidden, like deprecation logging is behind DeprecationLogger. But with header warnings, we already have the HeaderWarning class, and the logger retrieved will be the same for all locations, so we could just do this statically? Using a logger for header warnings directly is odd because the log level doesn't actual matter?